### PR TITLE
Add gc test for data virtualization

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
[AB#7412](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/7412)

Add gc tests for data virtualization.

The scenario here is the one I was most concerned about, basically does GC data passed down from the summary still work. Since we designed GC to work in "unrealized" datastores. The answer is yes, as this test proves such.

While I was testing this, I discovered that the call to `FluidDataStoreContext.isRoot` occurs during summarization for all datastores including virtualized datastores. This means that our "virtualization" would be moot.

I introduced a new change that would make it so that our virtualized datastores would not be loaded during summarization.

### CODE REVIEW

File: packages/runtime/container-runtime/src/channelCollection.ts [hunk 1]
```diff
@@ -1246,8 +1246,10 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
 	 */
 	private async getOutboundRoutes(): Promise<string[]> {
 		const outboundRoutes: string[] = [];
+		// Getting this information is a performance optimization that reduces network calls for virtualized datastores
+		const aliasedDataStores = new Set(this.aliasMap.values());
 		for (const [contextId, context] of this.contexts) {
-			const isRootDataStore = await context.isRoot();
+			const isRootDataStore = await context.isRoot(aliasedDataStores);
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/channelCollection.ts:1252]: Another thought here is that I could just do the check before running `await context.isRoot()`. Let me know what you prefer :)

File: packages/test/test-utils/src/testContainerRuntimeFactory.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ import {
 	IRequest,
 	IResponse,
 } from "@fluidframework/core-interfaces";
-// eslint-disable-next-line import/no-deprecated
```
Comment (tyler-cai-microsoft) [inline packages/test/test-utils/src/testContainerRuntimeFactory.ts:20]: I don't know why my auto save removed this line. But it did.

File: packages/test/test-utils/src/testContainerRuntimeFactory.ts [hunk 3]
```diff
@@ -89,7 +89,7 @@ export const createTestContainerRuntimeFactory = (
 			const rootContext =
 				"createDetachedRootDataStore" in runtime
 					? (runtime as any).createDetachedRootDataStore([this.type], "default")
-					: runtime.createDetachedDataStore([this.type], "default");
```
Comment (tyler-cai-microsoft) [inline packages/test/test-utils/src/testContainerRuntimeFactory.ts:92]: So this would create a "default" loadingGroup datastore.

File: packages/runtime/container-runtime/src/channelCollection.ts [hunk 4]
```diff
@@ -1246,8 +1246,10 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
 	 */
 	private async getOutboundRoutes(): Promise<string[]> {
 		const outboundRoutes: string[] = [];
+		// Getting this information is a performance optimization that reduces network calls for virtualized datastores
+		const aliasedDataStores = new Set(this.aliasMap.values());
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/channelCollection.ts:1250]: wondering if it's worth doing this once in the constructor, then add to this set each time we add to the alias map. This way, we don't have to incur an O(n) cost every time we need to call this function, which is every time GC runs.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDataVirtualization.spec.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+) => {
+	const interceptFn = async (...args: any[]) => {
+		const val = await fn.apply(parent, args);
+		intercept(val);
+		return val as T;
+	};
+	parent[fn.name] = interceptFn;
+	interceptFn.bind(parent);
+	return fn;
+};
```
Comment (andre4i) [inline packages/test/test-end-to-end-tests/src/test/gc/gcDataVirtualization.spec.ts:42]: By the 'Rule of three', this should probably go into some sort of 'common' file. Not right now, I'd say this is orthogonal to the PR.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDataVirtualization.spec.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+	async function isDataStoreInSummaryTree(summaryTree: ISummaryTree, dataStoreId: string) {
+		return getDataStoreInSummaryTree(summaryTree, dataStoreId) !== undefined;
+	}
+
+	beforeEach("setup", async function () {
+		provider = getTestObjectProvider({ syncSummarizer: true });
+
+		// Data virtualization only works with local
+		// TODO: enable for ODSP
```
Comment (andre4i) [inline packages/test/test-end-to-end-tests/src/test/gc/gcDataVirtualization.spec.ts:100]: Is there an ADO tracking this? With the risk of nitpicking, this should probably call out the ADO# here or remove the TODO altogether.

File: packages/test/test-utils/src/testContainerRuntimeFactory.ts (hunk 3 above)
Comment (andre4i) [inline packages/test/test-utils/src/testContainerRuntimeFactory.ts:92]: Why this change?

**Review by andre4i — APPROVED** [review]

File: packages/test/test-end-to-end-tests/src/test/gc/gcDataVirtualization.spec.ts (hunk 5 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/gc/gcDataVirtualization.spec.ts:42]: Agreed, I just haven't gotten around to it. Not sure which common file is recommended, I'll figure it out.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDataVirtualization.spec.ts (hunk 6 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/gc/gcDataVirtualization.spec.ts]: Updated
